File: sdks/go/aether/resilience/executor.go
package resilience

import (
	"context"
	"time"
)

// ResilientExecutor combines multiple resilience patterns.
//
// Order of operations:
// 1. Rate limiting (check if request is allowed)
// 2. Bulkhead (check capacity)
// 3. Circuit breaker (check if service is healthy)
// 4. Retry (handle transient failures)
type ResilientExecutor struct {
	breaker     *CircuitBreaker
	retry       *RetryPolicy
	rateLimiter *RateLimiter
	bulkhead    *Bulkhead
}

// ExecutorOption is a functional option for configuring the executor.
type ExecutorOption func(*ResilientExecutor)

// WithCircuitBreaker adds a circuit breaker to the executor.
func WithCircuitBreaker(breaker *CircuitBreaker) ExecutorOption {
	return func(e *ResilientExecutor) {
		e.breaker = breaker
	}
}

// WithRetry adds a retry policy to the executor.
func WithRetry(retry *RetryPolicy) ExecutorOption {
	return func(e *ResilientExecutor) {
		e.retry = retry
	}
}

// WithRateLimiter adds a rate limiter to the executor.
func WithRateLimiter(limiter *RateLimiter) ExecutorOption {
	return func(e *ResilientExecutor) {
		e.rateLimiter = limiter
	}
}

// WithBulkhead adds a bulkhead to the executor.
func WithBulkhead(bulkhead *Bulkhead) ExecutorOption {
	return func(e *ResilientExecutor) {
		e.bulkhead = bulkhead
	}
}

// NewResilientExecutor creates a new executor with the given options.
func NewResilientExecutor(opts ...ExecutorOption) *ResilientExecutor {
	e := &ResilientExecutor{}
	for _, opt := range opts {
		opt(e)
	}
	return e
}

// Execute runs the function with all configured resilience patterns.
func (e *ResilientExecutor) Execute(ctx context.Context, fn func(ctx context.Context) (any, error)) (any, error) {
	// Apply rate limiting first
	if e.rateLimiter != nil {
		if err := e.rateLimiter.Acquire(ctx, 5*time.Second); err != nil {
			return nil, err
		}
	}

	// Apply bulkhead
	if e.bulkhead != nil {
		return e.bulkhead.Execute(ctx, func(ctx context.Context) (any, error) {
			return e.executeWithRetry(ctx, fn)
		})
	}

	return e.executeWithRetry(ctx, fn)
}

// executeWithRetry applies retry logic.
func (e *ResilientExecutor) executeWithRetry(ctx context.Context, fn func(ctx context.Context) (any, error)) (any, error) {
	// Apply circuit breaker
	if e.breaker != nil {
		return e.breaker.Execute(ctx, func(ctx context.Context) (any, error) {
			return e.executeWithRetryInternal(ctx, fn)
		})
	}

	return e.executeWithRetryInternal(ctx, fn)
}

// executeWithRetryInternal applies retry without circuit breaker.
func (e *ResilientExecutor) executeWithRetryInternal(ctx context.Context, fn func(ctx context.Context) (any, error)) (any, error) {
	if e.retry != nil {
		result, err := e.retry.Execute(ctx, fn)
		if err != nil {
			return nil, err
		}
		return result, nil
	}

	return fn(ctx)
}

// ============================================
// Executor Builder
// ============================================

// ExecutorBuilder provides a fluent interface for building executors.
type ExecutorBuilder struct {
	breaker     *CircuitBreaker
	retry       *RetryPolicy
	rateLimiter *RateLimiter
	bulkhead    *Bulkhead
}

// NewExecutorBuilder creates a new builder.
func NewExecutorBuilder() *ExecutorBuilder {
	return &ExecutorBuilder{}
}

// CircuitBreaker sets the circuit breaker.
func (b *ExecutorBuilder) CircuitBreaker(config CircuitBreakerConfig) *ExecutorBuilder {
	b.breaker = NewCircuitBreaker(config)
	return b
}

// Retry sets the retry policy.
func (b *ExecutorBuilder) Retry(config RetryConfig) *ExecutorBuilder {
	b.retry = NewRetryPolicy(config)
	return b
}

// RateLimiter sets the rate limiter.
func (b *ExecutorBuilder) RateLimiter(config RateLimitConfig) *ExecutorBuilder {
	b.rateLimiter = NewRateLimiter(config)
	return b
}

// Bulkhead sets the bulkhead.
func (b *ExecutorBuilder) Bulkhead(config BulkheadConfig) *ExecutorBuilder {
	b.bulkhead = NewBulkhead(config)
	return b
}

// Build creates the executor.
func (b *ExecutorBuilder) Build() *ResilientExecutor {
	return NewResilientExecutor(
		WithCircuitBreaker(b.breaker),
		WithRetry(b.retry),
		WithRateLimiter(b.rateLimiter),
		WithBulkhead(b.bulkhead),
	)
}

// ============================================
// Predefined Executors
// ============================================

// DefaultExecutor creates an executor with sensible defaults.
func DefaultExecutor() *ResilientExecutor {
	return NewExecutorBuilder().
		CircuitBreaker(DefaultCircuitBreakerConfig()).
		Retry(DefaultRetryConfig()).
		RateLimiter(DefaultRateLimitConfig()).
		Bulkhead(DefaultBulkheadConfig()).
		Build()
}

// APIExecutor creates an executor for API calls.
func APIExecutor() *ResilientExecutor {
	return NewExecutorBuilder().
		CircuitBreaker(CircuitBreakerConfig{
			FailureThreshold: 5,
			SuccessThreshold: 3,
			Timeout:          30 * time.Second,
		}).
		Retry(RetryConfig{
			MaxAttempts:  3,
			Backoff:      BackoffExponentialJitter,
			BaseDelay:    100 * time.Millisecond,
			MaxDelay:     5 * time.Second,
			Multiplier:   2.0,
			JitterFactor: 0.1,
		}).
		RateLimiter(RateLimitConfig{
			RequestsPerSecond: 100,
			BurstSize:         200,
			Strategy:          StrategyTokenBucket,
		}).
		Bulkhead(BulkheadConfig{
			MaxConcurrent: 50,
			MaxQueued:     100,
		}).
		Build()
}

// DatabaseExecutor creates an executor for database operations.
func DatabaseExecutor() *ResilientExecutor {
	return NewExecutorBuilder().
		CircuitBreaker(CircuitBreakerConfig{
			FailureThreshold: 3,
			SuccessThreshold: 2,
			Timeout:          60 * time.Second,
		}).
		Retry(RetryConfig{
			MaxAttempts: 5,
			Backoff:     BackoffExponential,
			BaseDelay:   50 * time.Millisecond,
			MaxDelay:    2 * time.Second,
			Multiplier:  2.0,
		}).
		RateLimiter(RateLimitConfig{
			RequestsPerSecond: 50,
			BurstSize:         100,
			Strategy:          StrategyTokenBucket,
		}).
		Bulkhead(BulkheadConfig{
			MaxConcurrent: 10,
			MaxQueued:     50,
			Timeout:       30 * time.Second,
		}).
		Build()
}

// CriticalExecutor creates an executor for critical operations.
func CriticalExecutor() *ResilientExecutor {
	return NewExecutorBuilder().
		CircuitBreaker(CircuitBreakerConfig{
			FailureThreshold: 2,
			SuccessThreshold: 5,
			Timeout:          10 * time.Second,
		}).
		Retry(RetryConfig{
			MaxAttempts:   3,
			Backoff:       BackoffExponential,
			BaseDelay:     200 * time.Millisecond,
			MaxDelay:      10 * time.Second,
			Multiplier:    2.0,
			JitterFactor:  0.05,
		}).
		RateLimiter(RateLimitConfig{
			RequestsPerSecond: 20,
			BurstSize:         30,
			Strategy:          StrategySlidingWindow,
		}).
		Bulkhead(BulkheadConfig{
			MaxConcurrent: 5,
			MaxQueued:     20,
			Timeout:       10 * time.Second,
		}).
		Build()
}

File: sdks/javascript/src/resilience/executor.ts
/**
 * Resilient Executor - Combines all resilience patterns.
 * @module aether/resilience/executor
 */

import { CircuitBreaker } from './circuit_breaker';
import { RetryPolicy } from './retry';
import { RateLimiter } from './rate_limiter';
import { Bulkhead } from './bulkhead';
import { AsyncFunction } from './types';
import { withTracing } from './tracing';

/**
 * Configuration for {@link ResilientExecutor}.
 */
export interface ResilientExecutorConfig {
  /** Optional circuit breaker instance. */
  breaker?: CircuitBreaker;
  /** Optional retry policy instance. */
  retry?: RetryPolicy;
  /** Optional rate limiter instance. */
  rateLimiter?: RateLimiter;
  /** Optional bulkhead instance. */
  bulkhead?: Bulkhead;
  /** Name for tracing spans (default: `'default'`). */
  name?: string;
}

/**
 * Combined resilience executor that applies all configured patterns.
 *
 * The executor chains resilience patterns in the following order:
 *
 * 1. **Rate limiting** — checks if the request is allowed.
 * 2. **Bulkhead** — checks concurrency capacity.
 * 3. **Circuit breaker** — checks if the downstream service is healthy.
 * 4. **Retry** — handles transient failures with backoff.
 *
 * Each pattern is optional; only the configured ones are applied.
 *
 * @example
 * ```typescript
 * const executor = new ResilientExecutor({
 *   breaker: new CircuitBreaker({ failureThreshold: 5 }),
 *   retry: new RetryPolicy({ maxAttempts: 3 }),
 *   rateLimiter: new RateLimiter({ maxRequests: 100 }),
 *   bulkhead: new Bulkhead({ maxConcurrent: 10 }),
 * });
 *
 * const result = await executor.execute(() => fetchData());
 * ```
 */
export class ResilientExecutor {
  private readonly breaker?: CircuitBreaker;
  private readonly retry?: RetryPolicy;
  private readonly rateLimiter?: RateLimiter;
  private readonly bulkhead?: Bulkhead;
  private readonly name: string;

  /**
   * Create a new ResilientExecutor.
   *
   * @param config - Configuration with optional resilience components.
   */
  constructor(config: ResilientExecutorConfig = {}) {
    this.breaker = config.breaker;
    this.retry = config.retry;
    this.rateLimiter = config.rateLimiter;
    this.bulkhead = config.bulkhead;
    this.name = config.name ?? 'default';
  }

  /**
   * Execute a function with all configured resilience patterns.
   *
   * @typeParam T - The return type of the function.
   * @param fn - Async function to execute.
   * @returns Result of the function.
   * @throws CircuitBreakerError   If the circuit is open.
   * @throws RetryExhaustedError   If all retries fail.
   * @throws RateLimitExhaustedError If rate limit is exceeded.
   * @throws BulkheadRejectedError If bulkhead is at capacity.
   * @throws Error                 Any error thrown by the function itself.
   */
  async execute<T>(fn: AsyncFunction<T>): Promise<T> {
    return withTracing(`resilient_executor.${this.name}.execute`, async (span) => {
      span.setAttribute('resilient_executor.name', this.name);
      span.setAttribute('resilient_executor.has_circuit_breaker', !!this.breaker);
      span.setAttribute('resilient_executor.has_retry', !!this.retry);
      span.setAttribute('resilient_executor.has_rate_limiter', !!this.rateLimiter);
      span.setAttribute('resilient_executor.has_bulkhead', !!this.bulkhead);

      // 1. Apply rate limiting first
      if (this.rateLimiter) {
        await this.rateLimiter.acquire();
        span.addEvent('rate_limit.passed');
      }

      // 2. Apply bulkhead
      if (this.bulkhead) {
        return this.bulkhead.execute(() => this.executeWithRetry(fn, span));
      }

      return this.executeWithRetry(fn, span);
    });
  }

  /**
   * Execute with retry and circuit breaker.
   *
   * @typeParam T - The return type.
   * @param fn   - The async function.
   * @param span - The tracing span.
   * @returns The function result.
   * @internal
   */
  private async executeWithRetry<T>(
    fn: AsyncFunction<T>,
    span: any
  ): Promise<T> {
    // 3. Apply circuit breaker
    if (this.breaker) {
      return this.breaker.execute(() => this.executeWithRetryInternal(fn, span));
    }

    return this.executeWithRetryInternal(fn, span);
  }

  /**
   * Execute with retry logic.
   *
   * @typeParam T - The return type.
   * @param fn   - The async function.
   * @param span - The tracing span.
   * @returns The function result.
   * @internal
   */
  private async executeWithRetryInternal<T>(
    fn: AsyncFunction<T>,
    span: any
  ): Promise<T> {
    // 4. Apply retry
    if (this.retry) {
      const result = await this.retry.execute(fn);
      if (result.success) {
        span.setAttribute('resilient_executor.result', 'success');
        return result.result!;
      }
      span.setAttribute('resilient_executor.result', 'failed');
      span.setAttribute('resilient_executor.attempts', result.attempts);
      throw result.error!;
    }

    const result = await fn();
    span.setAttribute('resilient_executor.result', 'success');
    return result;
  }

  /**
   * Create a builder for fluently constructing a ResilientExecutor.
   *
   * @returns A new {@link ResilientExecutorBuilder}.
   */
  static builder(): ResilientExecutorBuilder {
    return new ResilientExecutorBuilder();
  }
}

/**
 * Fluent builder for constructing a {@link ResilientExecutor}.
 *
 * Each `with*` method returns `this` for chaining.
 *
 * @example
 * ```typescript
 * const executor = ResilientExecutor.builder()
 *   .withCircuitBreaker({ failureThreshold: 5 })
 *   .withRetry({ maxAttempts: 3 })
 *   .withRateLimiter({ maxRequests: 100 })
 *   .withBulkhead({ maxConcurrent: 10 })
 *   .withName('api-calls')
 *   .build();
 * ```
 */
export class ResilientExecutorBuilder {
  private config: ResilientExecutorConfig = {};

  /**
   * Add a circuit breaker with the given configuration.
   *
   * @param config - Partial circuit breaker configuration.
   * @returns This builder for chaining.
   */
  withCircuitBreaker(config?: ConstructorParameters<typeof CircuitBreaker>[0]): this {
    this.config.breaker = new CircuitBreaker(config ?? {});
    return this;
  }

  /**
   * Add an existing circuit breaker instance.
   *
   * @param breaker - The CircuitBreaker to use.
   * @returns This builder for chaining.
   */
  withExistingCircuitBreaker(breaker: CircuitBreaker): this {
    this.config.breaker = breaker;
    return this;
  }

  /**
   * Add a retry policy with the given configuration.
   *
   * @param config - Partial retry policy configuration.
   * @returns This builder for chaining.
   */
  withRetry(config?: ConstructorParameters<typeof RetryPolicy>[0]): this {
    this.config.retry = new RetryPolicy(config ?? {});
    return this;
  }

  /**
   * Add an existing retry policy instance.
   *
   * @param retry - The RetryPolicy to use.
   * @returns This builder for chaining.
   */
  withExistingRetry(retry: RetryPolicy): this {
    this.config.retry = retry;
    return this;
  }

  /**
   * Add a rate limiter with the given configuration.
   *
   * @param config - Partial rate limiter configuration.
   * @returns This builder for chaining.
   */
  withRateLimiter(config?: ConstructorParameters<typeof RateLimiter>[0]): this {
    this.config.rateLimiter = new RateLimiter(config ?? {});
    return this;
  }

  /**
   * Add an existing rate limiter instance.
   *
   * @param rateLimiter - The RateLimiter to use.
   * @returns This builder for chaining.
   */
  withExistingRateLimiter(rateLimiter: RateLimiter): this {
    this.config.rateLimiter = rateLimiter;
    return this;
  }

  /**
   * Add a bulkhead with the given configuration.
   *
   * @param config - Partial bulkhead configuration.
   * @returns This builder for chaining.
   */
  withBulkhead(config?: ConstructorParameters<typeof Bulkhead>[0]): this {
    this.config.bulkhead = new Bulkhead(config ?? {});
    return this;
  }

  /**
   * Add an existing bulkhead instance.
   *
   * @param bulkhead - The Bulkhead to use.
   * @returns This builder for chaining.
   */
  withExistingBulkhead(bulkhead: Bulkhead): this {
    this.config.bulkhead = bulkhead;
    return this;
  }

  /**
   * Set the name for tracing spans.
   *
   * @param name - The executor name.
   * @returns This builder for chaining.
   */
  withName(name: string): this {
    this.config.name = name;
    return this;
  }

  /**
   * Build and return the configured ResilientExecutor.
   *
   * @returns A new ResilientExecutor with the configured resilience patterns.
   */
  build(): ResilientExecutor {
    return new ResilientExecutor(this.config);
  }
}
